### PR TITLE
Add controller icons

### PR DESCRIPTION
**Pull request by TrojanerHD**
Uses https://github.com/rsubtil/controller_icons/ to implement controller icons / pretty keyboard icons

Picture of the F2 screen

Picture of a hint text for setting the reverser

Picture of a hint text for deaccelerating and accelerating

Picture of a hint text for opening doors at a station

### CODE REVIEW

**Comment by HaSa1002** [comment]
Here's a diff using a rich text label. See https://docs.godotengine.org/en/3.5/tutorials/ui/bbcode_in_richtextlabel.html#image-vertical-offset for an explanation why the bitmap font is required. Please reformat the addon scripts to tab based indention. Also, remove the example folder and copy the LICENSE into the addon folder, please. For some reason, they fail to detect my XSERIES X controller as such, but that's not your fault.
```diff
 src/Data/Fonts/image_offset_pseudo.tres       |  4 +++
 src/Data/UI/TrainInfo.tscn                    | 19 ++++++------
 src/Data/UI/input_rich_text_label.gd          | 30 +++++++++++++++++++
 .../controller_icons/ControllerIcons.gd       | 13 ++++++++
 src/project.godot                             |  6 ++++
 5 files changed, 63 insertions(+), 9 deletions(-)
 create mode 100644 src/Data/Fonts/image_offset_pseudo.tres
 create mode 100644 src/Data/UI/input_rich_text_label.gd

diff --git a/src/Data/Fonts/image_offset_pseudo.tres b/src/Data/Fonts/image_offset_pseudo.tres
new file mode 100644
index 0000000..57fdeac
--- /dev/null
+++ b/src/Data/Fonts/image_offset_pseudo.tres
@@ -0,0 +1,4 @@
+[gd_resource type="BitmapFont" format=2]
+
+[resource]
+ascent = 6.0
diff --git a/src/Data/UI/TrainInfo.tscn b/src/Data/UI/TrainInfo.tscn
index e7cff69..9e27780 100644
--- a/src/Data/UI/TrainInfo.tscn
+++ b/src/Data/UI/TrainInfo.tscn
@@ -1,4 +1,4 @@
-[gd_scene load_steps=9 format=2]
+[gd_scene load_steps=10 format=2]
 
 [ext_resource path="res://Data/Fonts/FontMedium.tres" type="DynamicFont" id=1]
 [ext_resource path="res://Data/Fonts/FontIngame.tres" type="DynamicFont" id=2]
@@ -7,6 +7,7 @@
 [ext_resource path="res://Data/Misc/DotGreen.png" type="Texture" id=5]
 [ext_resource path="res://Data/Scripts/InputLabel.gd" type="Script" id=6]
 [ext_resource path="res://addons/controller_icons/objects/TextureRect.gd" type="Script" id=7]
+[ext_resource path="res://Data/UI/input_rich_text_label.gd" type="Script" id=8]
 [ext_resource path="res://addons/controller_icons/assets/key/b.png" type="Texture" id=9]
 
 [node name="TrainInfo" type="Control"]
@@ -129,16 +130,16 @@ margin_right = 30.0
 margin_bottom = 30.0
 texture = ExtResource( 3 )
 
-[node name="Label" type="Label" parent="ScrollContainer/VBoxContainer/Engine"]
+[node name="Label" type="RichTextLabel" parent="ScrollContainer/VBoxContainer/Engine"]
 margin_left = 34.0
-margin_right = 181.0
-margin_bottom = 30.0
-custom_fonts/font = ExtResource( 2 )
+margin_right = 753.0
+margin_bottom = 31.0
+size_flags_horizontal = 3
+custom_fonts/normal_font = ExtResource( 2 )
 text = "TRAIN_INFO_ENGINE"
-script = ExtResource( 6 )
-__meta__ = {
-"_edit_use_anchors_": false
-}
+fit_content_height = true
+scroll_active = false
+script = ExtResource( 8 )
 actions = [ "engine" ]
 
 [node name="EnforcedBreaking" type="HBoxContainer" parent="ScrollContainer/VBoxContainer"]
diff --git a/src/Data/UI/input_rich_text_label.gd b/src/Data/UI/input_rich_text_label.gd
new file mode 100644
index 0000000..ca8b44e
--- /dev/null
+++ b/src/Data/UI/input_rich_text_label.gd
@@ -0,0 +1,30 @@
+class_name InputRichTextLabel
+extends RichTextLabel
+
+
+export(Array, String) var actions := []
+
+
+onready var translation_id := text
+
+
+func _ready() -> void:
+	bbcode_enabled = true
+	update_text()
+	ControllerIcons.connect("input_type_changed", self, "update_text")
+
+
+func _notification(what: int) -> void:
+	if what == NOTIFICATION_TRANSLATION_CHANGED:
+		update_text()
+
+
+func set_text(text: String) -> void:
+	.set_text(tr(text))
+
+
+func update_text(var _x = null) -> void:
+	var replaces := []
+	for action in actions:
+		replaces.push_back("[font=res://Data/Fonts/image_offset_pseudo.tres][img=36]%s[/img][/font]" % ControllerIcons.get_action_path(action))
+	bbcode_text = tr(translation_id) % replaces
diff --git a/src/addons/controller_icons/ControllerIcons.gd b/src/addons/controller_icons/ControllerIcons.gd
index 66fba40..756ef30 100644
--- a/src/addons/controller_icons/ControllerIcons.gd
+++ b/src/addons/controller_icons/ControllerIcons.gd
@@ -100,6 +100,19 @@ func parse_path(path: String, input_type: int = _last_input_type) -> Texture:
     return _cached_icons[root_path]
   return null
 
+
+func get_action_path(path: String, input_type: int = _last_input_type) -> String:
+	if input_type == null:
+		return ""
+	var root_paths := _expand_path(path, input_type)
+	for root_path in root_paths:
+		if not _cached_icons.has(root_path):
+			if _load_icon(root_path):
+				continue
+		return root_path
+	return ""
+
+
 func parse_event(event: InputEvent) -> Texture:
   var path = _convert_event_to_path(event)
   if path.empty():
diff --git a/src/project.godot b/src/project.godot
index c7d5588..5e2f84a 100644
--- a/src/project.godot
+++ b/src/project.godot
@@ -139,6 +139,11 @@ _global_script_classes=[ {
 "language": "GDScript",
 "path": "res://addons/jean28518.jTools/jSettings/input_map_resource.gd"
 }, {
+"base": "RichTextLabel",
+"class": "InputRichTextLabel",
+"language": "GDScript",
+"path": "res://Data/UI/input_rich_text_label.gd"
+}, {
 "base": "PanelContainer",
 "class": "InputSettings",
 "language": "GDScript",
@@ -386,6 +391,7 @@ _global_script_class_icons={
 "InputButton": "",
 "InputLabel": "",
 "InputMapResource": "",
+"InputRichTextLabel": "",
 "InputSettings": "",
 "LTSPlayer": "",
 "LTSWorld": "",
```

**Comment by TrojanerHD** [comment]
I think I implemented everything the way you requested now and updated the original PR post
However, I would say that the texts in the F2 screen would probably look better without the quotation marks. Thoughts?

EDIT: I also was not sure about how to handle multiple inputs (like CTRL + L for cabin light). Would it look better with or without a `+` inbetween?

**Comment by HaSa1002** [comment]
Yeah, I'd say kill them (all). Leave them without a `+` for now. I think that's fine.

**Comment by HaSa1002** [comment]
Yeah, I'd say kill them (all). Leave them without a `+` for now. I think that's fine.

File: src/Data/Scripts/HUD.gd [hunk 1]
```diff
@@ -38,10 +38,10 @@ func _unhandled_input(_event) -> void:
 
 var messages: int = 0
 func send_message(text: String, actions := []) -> void:
-	var Message: InputLabel = $PanelContainer/MessageLabel as InputLabel
-	Message.backing_text = text
+	var Message: RichTextLabel = $PanelContainer/MessageLabel as RichTextLabel
```
Comment (HaSa1002) [inline src/Data/Scripts/HUD.gd]:
```suggestion
	var Message  := $PanelContainer/MessageLabel as InputRichTextLabel
```

Please use the specific scrpt class whenever possible

File: src/Data/Scripts/TrainInfo.gd [hunk 2]
```diff
@@ -29,27 +29,15 @@ func update_info(player: LTSPlayer) -> void:
...
 		else:
 			$ScrollContainer/VBoxContainer/Doors/dot.texture = red
-
-	## Control Type:
-	if player.control_type == player.ControlType.COMBINED:
-		$"ScrollContainer/VBoxContainer/Brakes-1".hide()
-		$"ScrollContainer/VBoxContainer/Acceleration-1".hide()
-	else:
-		$"ScrollContainer/VBoxContainer/Brakes-0".hide()
-		$"ScrollContainer/VBoxContainer/Acceleration-0".hide()
```
Comment (HaSa1002) [inline src/Data/Scripts/TrainInfo.gd]: Y'know these are two distinct input modes? How are they handled now?

File: src/Data/UI/input_rich_text_label.gd [hunk 3]
```diff
@@ -0,0 +1,37 @@
+class_name InputRichTextLabel
+extends RichTextLabel
+
+
+export(Array, String) var actions := []
+export(bool) var centered := false
```
Comment (HaSa1002) [inline src/Data/UI/input_rich_text_label.gd]:
```suggestion
export var centered := false
```

Type is inferred. I only used (Array, String) above to give a better hint than a generic array

File: src/Data/UI/input_rich_text_label.gd [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+	var replaces := []
+	for possible_action in actions:
+		var combinations := ""
+		for action in ControllerIcons.get_action_paths(possible_action):
+			combinations += "[font=res://Data/Fonts/image_offset_pseudo.tres][img=36]%s[/img][/font]" % action
+		replaces.push_back(combinations)
+	bbcode_text = ""
+	if centered:
+		bbcode_text = "[center]"
+	append_bbcode(tr(translation_id) % replaces)
```
Comment (HaSa1002) [inline src/Data/UI/input_rich_text_label.gd]:
```suggestion
	if centered:
		bbcode_text = "[center]%s[/center]" % (tr(translation_id) % replaces)
		return
	bbcode_text = tr(translation_id) % replaces
```

You forgot a `/center` there. Also, no need to update the RTL content twice when assigning the text like this

File: src/project.godot [hunk 5]
```diff
@@ -446,7 +489,7 @@ scene_naming=2
 
 [editor_plugins]
 
-enabled=PoolStringArray( "res://addons/HaSa1002.light-tools/plugin.cfg", "res://addons/libre_trainsim_modding_tools/plugin.cfg" )
+enabled=PoolStringArray( "res://addons/libre_trainsim_modding_tools/plugin.cfg" )
```
Comment (HaSa1002) [inline src/project.godot]: Please revert that change.

File: src/project.godot [hunk 6]
```diff
@@ -771,6 +814,10 @@ translation_remaps={
 overwritten_language=""
 overwritten_language.editor="en"
 
+[mono]
+
+project/assembly_name="Libre Train Sim"
+
```
Comment (HaSa1002) [inline src/project.godot]: Please revert that, too. We run without C#

**Review by HaSa1002 — CHANGES_REQUESTED** [review]
Haven't tested yet. I just did a quick code inspection of code and text assets. A test follows

File: src/Data/Scripts/TrainInfo.gd (hunk 2 above)
Comment (TrojanerHD) [reply src/Data/Scripts/TrainInfo.gd]: Oh whoops, I misinterpreted the code then. What exactly are the input modes?

Comment (HaSa1002) [reply src/Data/Scripts/TrainInfo.gd]: We have combined mode where you control the train only with up and down. So brake and acceleration are on the same axis.
In the other one there you have acceleration on up and down and the brake on left and right. The idea being to simulate the two distinct controls (control wheels). We default to shared mode but if people deselect easy mode, they have split controls.

File: src/addons/controller_icons/ControllerIcons.gd [hunk 7]
```diff
@@ -0,0 +1,523 @@
...
+	if input_type == null:
+		return []
+	var root_paths := _expand_path(path, input_type)
+	var event = _get_matching_event(path, input_type)
+	
+	for root_path in root_paths:
+		if not _cached_icons.has(root_path):
+			if _load_icon(root_path):
+				continue
+		var combination := [root_path]
```
Comment (HaSa1002) [inline src/addons/controller_icons/ControllerIcons.gd:132]:
```suggestion
		var combination := [root_path]
		var mk := event as InputEventWithModifiers
		if mk == null:
			return combination
```

It fails if the input comes from a controller.

Inluded in patch

File: src/addons/controller_icons/settings.tres [hunk 8]
```diff
@@ -0,0 +1,11 @@
+[gd_resource type="Resource" load_steps=2 format=2]
+
+[ext_resource path="res://addons/controller_icons/Settings.gd" type="Script" id=1]
+
+[resource]
+script = ExtResource( 1 )
+joypad_fallback = 12
```
Comment (HaSa1002) [inline src/addons/controller_icons/settings.tres]:
```suggestion
joypad_fallback = 14
```

Let's use Xbox Series X controller icons as fallback. These are less colourful and sadly on Windows we have no chance to distinguish what Xbox controller that is.

Included in patch

File: src/Data/UI/HUD.tscn [hunk 9]
```diff
@@ -280,22 +282,20 @@ anchor_top = 1.0
...
+[node name="MessageLabel" type="RichTextLabel" parent="PanelContainer"]
 margin_left = 7.0
 margin_top = 7.0
 margin_right = 1017.0
-margin_bottom = 52.0
-custom_fonts/font = ExtResource( 16 )
+margin_bottom = 38.0
+size_flags_horizontal = 13
+size_flags_vertical = 4
+custom_fonts/normal_font = ExtResource( 19 )
```
Comment (HaSa1002) [inline src/Data/UI/HUD.tscn]:
```suggestion
custom_fonts/normal_font = ExtResource( 16 )
```

Please stick with the big font. You may need to make the image_offset_pseudo font configurable with the existing one being the default.

**Review by HaSa1002 — CHANGES_REQUESTED** [review]
Looking good. I made you another patch of stuff I found to be fixed. Basically the addon only checked for the first controller and not the controller the input came from. In this patch are also the changes in `settings.tres` and the fix in `ControllerIcons.gd` regarding the event

```diff
diff --git a/src/addons/controller_icons/ControllerIcons.gd b/src/addons/controller_icons/ControllerIcons.gd
index 6f00361..bdc3b7b 100644
--- a/src/addons/controller_icons/ControllerIcons.gd
+++ b/src/addons/controller_icons/ControllerIcons.gd
@@ -12,6 +12,7 @@ var _cached_icons := {}
 var _custom_input_actions := {}
 
 var _last_input_type
+var _last_controller := -1
 var _settings
 var _file := File.new()
 
@@ -68,6 +69,7 @@ func _on_joy_connection_changed(device, connected):
 
 func _input(event: InputEvent):
 	var input_type = _last_input_type
+	var controller = _last_controller
 	match event.get_class():
 		"InputEventKey", "InputEventMouseButton":
 			input_type = InputType.KEYBOARD_MOUSE
@@ -76,10 +78,14 @@ func _input(event: InputEvent):
 				input_type = InputType.KEYBOARD_MOUSE
 		"InputEventJoypadButton":
 			input_type = InputType.CONTROLLER
+			controller = event.device
 		"InputEventJoypadMotion":
 			if abs(event.axis_value) > _settings.joypad_deadzone:
 				input_type = InputType.CONTROLLER
-	if input_type != _last_input_type:
+				controller = event.device
+	if input_type != _last_input_type or \
+			(input_type == InputType.CONTROLLER and _last_controller != controller):
+		_last_controller = controller
 		_set_last_input_type(input_type)
 
 func _add_custom_input_action(input_action: String, data: Dictionary):
@@ -118,19 +124,22 @@ func get_action_paths(path: String, input_type: int = _last_input_type) -> Array
 		return []
 	var root_paths := _expand_path(path, input_type)
 	var event = _get_matching_event(path, input_type)
-	
+
 	for root_path in root_paths:
 		if not _cached_icons.has(root_path):
 			if _load_icon(root_path):
 				continue
 		var combination := [root_path]
-		if event.control:
+		var mk := event as InputEventWithModifiers
+		if mk == null:
+			return combination
+		if mk.control:
 			combination.push_front(fake_event(KEY_CONTROL))
-		if event.alt:
+		if mk.alt:
 			combination.push_front(fake_event(KEY_ALT))
-		if event.shift:
+		if mk.shift:
 			combination.push_front(fake_event(KEY_SHIFT))
-		if event.meta:
+		if mk.meta:
 			combination.push_front(fake_event(KEY_META))
 		return combination
 
@@ -192,7 +201,7 @@ func _expand_path(path: String, input_type: int) -> Array:
 			if event:
 				base_path += _convert_event_to_path(event)
 		elif path.substr(0, path.find("/")) == "joypad":
-			base_path += Mapper._convert_joypad_path(path, _settings.joypad_fallback)
+			base_path += Mapper._convert_joypad_path(path, _settings.joypad_fallback, _last_controller)
 		else:
 			base_path += path
 
@@ -466,7 +475,7 @@ func _convert_joypad_button_to_path(button_index: int):
 			path = "joypad/share"
 		_:
 			return ""
-	return Mapper._convert_joypad_path(path, _settings.joypad_fallback)
+	return Mapper._convert_joypad_path(path, _settings.joypad_fallback, _last_controller)
 
 func _convert_joypad_motion_to_path(axis: int):
 	var path : String
@@ -481,7 +490,7 @@ func _convert_joypad_motion_to_path(axis: int):
 			path = "joypad/rt"
 		_:
 			return ""
-	return Mapper._convert_joypad_path(path, _settings.joypad_fallback)
+	return Mapper._convert_joypad_path(path, _settings.joypad_fallback, _last_controller)
 
 func _get_matching_event(path: String, input_type: int):
 	var events : Array
diff --git a/src/addons/controller_icons/Mapper.gd b/src/addons/controller_icons/Mapper.gd
index 97f7ffd..7831889 100644
--- a/src/addons/controller_icons/Mapper.gd
+++ b/src/addons/controller_icons/Mapper.gd
@@ -1,8 +1,8 @@
 extends Node
 class_name ControllerMapper
 
-func _convert_joypad_path(path: String, fallback) -> String:
-	match _get_joypad_type(fallback):
+func _convert_joypad_path(path: String, fallback, controller_id) -> String:
+	match _get_joypad_type(fallback, controller_id):
 		ControllerSettings.Devices.LUNA:
 			return _convert_joypad_to_luna(path)
 		ControllerSettings.Devices.PS3:
@@ -30,8 +30,9 @@ func _convert_joypad_path(path: String, fallback) -> String:
 		_:
 			return ""
 
-func _get_joypad_type(fallback):
-	var controller_name = Input.get_joy_name(0)
+func _get_joypad_type(fallback, controller_id):
+	var controller_name = Input.get_joy_name(controller_id)
+	printt(controller_id, Input.get_joy_guid(controller_id), controller_name)
 	if "Luna Controller" in controller_name:
 		return ControllerSettings.Devices.LUNA
 	elif "PS3 Controller" in controller_name:
@@ -297,4 +298,4 @@ func _convert_joypad_to_steamdeck(path: String):
 		"share":
 			return path.replace("/share", "/dots")
 		_:
-			return path
\ No newline at end of file
+			return path
diff --git a/src/addons/controller_icons/settings.tres b/src/addons/controller_icons/settings.tres
index 93e98c5..a985b1f 100644
--- a/src/addons/controller_icons/settings.tres
+++ b/src/addons/controller_icons/settings.tres
@@ -4,7 +4,7 @@
 
 [resource]
 script = ExtResource( 1 )
-joypad_fallback = 12
+joypad_fallback = 14
 joypad_deadzone = 0.5
 allow_mouse_remap = true
 mouse_min_movement = 200
```

File: src/Data/UI/HUD.tscn (hunk 9 above)
Comment (TrojanerHD) [reply src/Data/UI/HUD.tscn]: I didn't find a better way than to hard code the medium font using a different ascend value by creating a separate pseudo font for the medium font. Is that fine or is there a better way?

Comment (HaSa1002) [reply src/Data/UI/HUD.tscn]: That's fine

File: src/Data/UI/input_rich_text_label.gd [hunk 10]
```diff
@@ -29,9 +29,15 @@ func update_text(var _x = null) -> void:
...
 	if centered:
 		bbcode_text = "[center]%s[/center]" % (tr(translation_id) % replaces)
 		return
 	bbcode_text = tr(translation_id) % replaces
+
+func get_font_specifier() -> String:
+	var font = get("custom_fonts/normal_font").resource_path.get_file()
+	if font == "FontMedium.tres":
+		return "medium"
+	return "ingame"
```
Comment (HaSa1002) [inline src/Data/UI/input_rich_text_label.gd:43]: That's a little bit much magic for my taste but we can fix if necessary

**Review by HaSa1002 — APPROVED** [review]
lgtm

**Comment by HaSa1002** [comment]
Do you wanna do a rebase and squash your commits or should I combine all o' them into a single commit? Your choice.

**Comment by TrojanerHD** [comment]
I would rather prefer that you combine them into one commit to keep my commit history ^^

**Comment by HaSa1002** [comment]
Congrats for your first merged PR! 🎉

**Pull request merged**